library/src/scala/Tuple.scala

```diff
@@ -232,6 +237,13 @@ object Tuple {
 
   def fromProductTyped[P <: Product](p: P)(using m: scala.deriving.Mirror.ProductOf[P]): m.MirroredElemTypes =
     runtime.Tuples.fromProduct(p).asInstanceOf[m.MirroredElemTypes]
+
+  // TODO: When we can break TASTy compat, move this method to `Tuple` class and use the following signature
+  //       `inline def toList[This >: this.type <: Tuple]: List[Union[This]]`
+  /** Create a copy this tuple as a List */
+  extension [This <: Tuple](inline tuple: This)
+    inline def toList: List[Union[This]] =
```

Comment:
What's the difference between this and List[tuple.type] ?

Reply:
It would fail with
```scala
3 |  (??? : Tuple).toList.map(bar)
  |  ^^^^^^^^^^^^^^^^^^^^
  |  Tuple is not a valid singleton type, since it is not an immutable path
  | This location contains code that was inlined from Tuple.scala:246
```
Also, the source of the original bug was caused by using the `this.type` instead of the type parameter.

Reply:
> Tuple is not a valid singleton type, since it is not an immutable path

I find it weird that using a singleton type in the method result type would produce this error, is this a bug in the interaction between match types and singleton types? /cc @OlivierBlanvillain

Reply:
I belive it is because the prefix is not a singleton type and hence we have no way to refer to `tuple.type`.

Reply:
Sounds like it's widening too early?  To me `Tuple` isn't the type that should be singleton, it's the value `(??? : Tuple)` that should be.  Put in two lines:

```scala
val foo = ??? : Tuple
foo.toList.map(bar)
```

I expect tuple.type=foo.type not Tuple.

Reply:
So what I'm trying to say is I think this PR is working around a problem.

Reply:
> the prefix is not a singleton type and hence we have no way to refer to tuple.type.

That's not a problem normally: if the prefix is not a singleton type and we need it to be, then we skolemize it, but I haven't thought about how that logic is supposed to interact with match types: https://github.com/lampepfl/dotty/blob/e59c5ea77e4a498cd3a196f6eb1fc6a6b4dab05e/compiler/src/dotty/tools/dotc/core/TypeOps.scala#L36-L43

Reply:
Then we are missing some skolemization for this case